src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,152 @@
+                            conditionVariable,
+                            lua.createBinaryExpression(conditionVariable, condition, lua.SyntaxKind.OrOperator)
+                        )
+                    );
+                    condition = undefined;
+                }
 
-        const goto = lua.createGotoStatement(`${switchName}_case_${index}`);
-        return lua.createIfStatement(condition, lua.createBlock([goto]), previousCondition);
-    }, undefined as lua.IfStatement | undefined);
+                statements.push(
+                    lua.createIfStatement(
+                        conditionVariable,
+                        lua.createBlock(context.transformStatements(clause.statements))
+                    )
+                );
+            }
 
-    if (concatenatedIf) {
-        statements.push(concatenatedIf);
-    }
+            previousClause = clause;
+        });
 
-    const hasDefaultCase = statement.caseBlock.clauses.some(ts.isDefaultClause);
-    statements.push(lua.createGotoStatement(`${switchName}_${hasDefaultCase ? "case_default" : "end"}`));
+        // Amalgamate the default w/ fallthrough clauses and execute if nothing else executed above
+        const start = clauses.findIndex(c => ts.isDefaultClause(c));
+        if (start >= 0) {
+            const end = statement.caseBlock.clauses
+                .slice(start)
+                .findIndex(clause => containsBreakOrReturn([...clause.statements]));
+            const defaultStatements = statement.caseBlock.clauses
+                .slice(start, end >= 0 ? end + start + 1 : undefined)
+                .reduce<lua.Statement[]>(
+                    (statements, clause) => [...statements, ...context.transformStatements(clause.statements)],
+                    []
+                );
 
-    for (const [index, clause] of statement.caseBlock.clauses.entries()) {
-        const labelName = `${switchName}_case_${ts.isCaseClause(clause) ? index : "default"}`;
-        statements.push(lua.createLabelStatement(labelName));
-        statements.push(lua.createDoStatement(context.transformStatements(clause.statements)));
+            if (defaultStatements.length) {
+                statements.push(
+                    lua.createIfStatement(
+                        lua.createUnaryExpression(conditionVariable, lua.SyntaxKind.NotOperator),
+                        lua.createBlock(defaultStatements)
+                    )
+                );
+            }
+        }
     }
 
-    statements.push(lua.createLabelStatement(`${switchName}_end`));
-
+    // Hoist the variable, function, and import statements to the top of the switch
     statements = performHoisting(context, statements);
     popScope(context);
 
+    // Add the switch expression after hoisting
     const expression = context.transformExpression(statement.expression);
+    statements.unshift(lua.createVariableDeclarationStatement(conditionVariable, lua.createBooleanLiteral(false)));
```

Comment:
Would it be possible to initialize the condition variable to the first case's expression?

Reply:
Sure. I was initially concerned that hoisting was going to mess with me. so opted for the most straightforward.
Also, it saves an if in the code above and allows for the same logic for each assignment. Going to clean up everything else first, then take a look at how much more complicated the code might look.